File: src/main/java/me/jameschan/burrow/kernel/command/chamber/HelpCommand.java
package me.jameschan.burrow.kernel.command.chamber;

import me.jameschan.burrow.kernel.command.Command;
import me.jameschan.burrow.kernel.common.ExitCode;
import me.jameschan.burrow.kernel.context.RequestContext;
import picocli.CommandLine;

@CommandLine.Command(name = "help", description = "Show the usage of a command.")
public class HelpCommand extends Command {
  @CommandLine.Parameters(
      index = "0",
      description = "The name of the command.",
      defaultValue = "help")
  private String commandName;

  public HelpCommand(final RequestContext context) {
    super(context);
  }

  @Override
  public Integer call() {
    final var command = context.getProcessor().getCommand(commandName);
    if (command == null) {
      buffer.append("No such command: ").append(commandName);
    } else {
      buffer.append(new CommandLine(command).getUsageMessage());
    }

    return ExitCode.SUCCESS;
  }
}

File: src/main/kotlin/carton/standard/command/HelpCommand.kt
package burrow.carton.standard.command

import burrow.kernel.command.Command
import burrow.kernel.command.CommandData
import picocli.CommandLine
import picocli.CommandLine.ExitCode

@CommandLine.Command(
    name = "help",
    description = ["Displays the usage of a specific command."]
)
class HelpCommand(data: CommandData) : Command(data) {
    @CommandLine.Parameters(
        index = "0",
        description = ["The name of the command."],
        defaultValue = ""
    )
    private var commandName: String = ""

    override fun call(): Int {
        return if (commandName.isEmpty())
            displayChamberUsage()
        else
            displayCommandUsage(commandName)
    }

    private fun displayChamberUsage(): Int {
        return ExitCode.OK
    }

    private fun displayCommandUsage(commandName: String): Int {
        val commandClass = processor.commandClasses[commandName]
        if (commandClass == null) {
            stderr.println("Unknown command $commandName")
            return ExitCode.USAGE
        }

        stdout.println(CommandLine(commandClass.java).usageMessage)
        return ExitCode.OK
    }
}